haskell-src/Concordium/Types/Transactions.hs

```diff
@@ -190,6 +191,138 @@ instance HashableTo TransactionHashV0 AccountTransaction where
+        let bitmap = S.putWord16be $ bitFor 0 thv1Sponsor
+            v0header = S.put thv1Sender <> S.put thv1Nonce <> S.put thv1EnergyAmount <> S.put thv1PayloadSize <> S.put thv1Expiry
+            optionals = maybe mempty S.put thv1Sponsor
+        in  bitmap <> v0header <> optionals
+      where
+        -- \| Set the given bit if the value is a 'Just'.
+        bitFor :: (Bits b) => Int -> Maybe a -> b
+        bitFor _ Nothing = zeroBits
+        bitFor i (Just _) = bit i
+
+    get = S.label "transaction header v1" $ do
+        bitmap <- S.label "bitmap" S.getWord16be
+        thv1Sender <- S.label "sender" S.get
+        thv1Nonce <- S.label "nonce" S.get
+        thv1EnergyAmount <- S.label "energy amount" S.get
+        thv1PayloadSize <- S.label "payload size" S.get
+        thv1Expiry <- S.label "expiry" S.get
+        thv1Sponsor <- maybeGet bitmap 0 "sponsor"
+        return $! TransactionHeaderV1{..}
+      where
+        maybeGet bitmap bitNum label
+            | testBit bitmap bitNum = Just <$> S.label label S.get
+            | otherwise = return Nothing
+
+-- | An 'AccountTransactionV1' is a transaction that originates from
+--  a specific account (the sender), and is paid for by either the sender
+--  or a sponsor account.
+--
+--  The representation includes a 'TransactionSignHash' which is
+--  the value that is signed. This is derived from the header and
+--  payload, and so does not form part of the serialization.
+--
+--  The payload is stored in serialized form. Deserializing the
+--  payload is considered part of the transaction execution.
+data AccountTransactionV1 = AccountTransactionV1
+    { -- | Signature
+      atrv1Signature :: !TransactionSignaturesV1,
+      -- | Header
+      atrv1Header :: !TransactionHeaderV1,
+      -- | Serialized payload
+      atrv1Payload :: !EncodedPayload,
+      -- | Hash used for signing
+      atrv1SignHash :: !TransactionSignHashV0
+    }
+    deriving (Eq, Show)
+
+v1TransactionSignHashPrefix :: BS.ByteString
+v1TransactionSignHashPrefix = BS.pack [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
+
+-- | Construct an 'AccountTransactionV1', computing the correct
+--  'TransactionSignHash'.
+makeAccountTransactionV1 :: TransactionSignaturesV1 -> TransactionHeaderV1 -> EncodedPayload -> AccountTransactionV1
+makeAccountTransactionV1 atrv1Signature atrv1Header atrv1Payload = AccountTransactionV1{..}
+  where
+    atrv1SignHash = transactionV1SignHashFromHeaderPayload atrv1Header atrv1Payload
+
+-- | Construct a 'TransactionSignHash' from a 'TransactionHeaderV1' and 'EncodedPayload'.
+transactionV1SignHashFromHeaderPayload :: TransactionHeaderV1 -> EncodedPayload -> TransactionSignHashV0
+transactionV1SignHashFromHeaderPayload atrv1Header atrv1Payload = transactionSignHashFromBytes $ v1TransactionSignHashPrefix <> bodyBytes
+  where
+    bodyBytes = S.runPut $ S.put atrv1Header <> putEncodedPayload atrv1Payload
```

Comment:
```suggestion
transactionV1SignHashFromHeaderPayload atrv1Header atrv1Payload = transactionSignHashFromBytes bytes
  where
    bytes = S.runPut $ do
        S.putByteString v1TransactionSignHashPrefix
        S.put atrv1Header
        putEncodedPayload atrv1Payload
```
It is more efficient to serialize the prefix in the putter than concatenate with it outside the putter. Concatenating strict byte strings is kind-of expensive, because it means allocating a new byte string and copying the existing ones. Within `runPut`, it's building the byte string more lazily, so it only copies when you actually construct it. In principle, I think it would be better to use a lazy byte string (cf. `transactionSignHashFromHeaderPayload`) as this (in theory) avoids copying altogether.